Support loading/using tools that are not in the install database

natefoo opened the pull request:
Note that such tools do not appear in the admin interface, but the presumed use case is for tools that you are not managing in this particular Galaxy instance (e.g. they are in CVMFS) anyway. The only downside is you won't be able to use the admin UI to view and fix dependencies (which we also expect to be installed in CVMFS, but you could imagine wanting to resolve them locally instead), but that might be fixable.

natefoo committed:
WIP load shed tools w/o install DB

natefoo committed:
Fix "show in toolshed" link

natefoo committed:
WIP broken support tool help w/out repository_id and start
de-hackifying

natefoo committed:
Fix generation of repo_path

natefoo committed:
Got tool help images working

natefoo committed:
Better condition handling for populating a tool's shed info

natefoo committed:
Fix silent failure when installing tools.

natefoo committed:
Fix installing TS tools

natefoo committed:
Lint

mvdbeek commented:
> admin UI to view and fix dependencies (which we also expect to be installed in CVMFS, but you could imagine wanting to resolve them locally instead), but that might be fixable.

That is only an issue for toolshed packages, so not a concern I think. For everything else there's the Manage dependencies menu, that lists everything.

mvdbeek commented on lib/tool_shed/tools/tool_validator.py:
```diff
@@ -291,16 +291,10 @@ def load_tool_from_config(self, repository_id, full_path):
-            valid = False
-            error_message = str(e)
+        except Exception:
+            # As best I can tell, error messages will no longer get back to the UI, so raise exceptions instead so that
```
That doesn't look right to me, there's some additional logic in the calling code that checks if a tool is valid and, if it makes sense, displays `error_message`. Otherwise no need to even return `valid` and `error_message`, I think.

natefoo replied:
So the situation I was encountering was an exception due to a change to the method signature of `Tool.__init__()`. The clone was succeeding, the tool was "installed" as far as the install DB was concerned, but it wasn't being added to `shed_tool_conf.xml`. There was no error reported in the UI or log file. So although the error was of my own making, figuring it out was a real nightmare. Presumably there are other non-development errors that would be missed by this as well?

mvdbeek replied:
I'm sure there are, but this breaks all code that uses the current logic. How about just logging the exception here?

natefoo commented:
> For everything else there's the Manage dependencies menu, that lists everything.

Awesome, I didn't know that feature didn't use the install DB.

natefoo committed:
Add unnecessary column

mvdbeek commented:
> Awesome, I didn't know that feature didn't use the install DB.

We should make this more prominent in the admin panel, I think it's the most frequent advice I send out to the mailing list / gitter. And it could use some UI love ...

natefoo committed:
Log exceptions instead of raising.

mvdbeek committed:
Some small fixes and improvements

Most notably only use tool_shed_repository while constructing a Tool,
this avoids stale references to sqlalchemy objects.

mvdbeek committed:
Fix display of images in static/images

mvdbeek committed:
Add dummy tool shed registry

mvdbeek committed:
Remove useless lock

mvdbeek committed:
Mock the mock shed registry

mvdbeek committed:
Fix set_image_path for tools in tool shed

mvdbeek committed:
Simplify getting changeset revision from path

mvdbeek committed:
Simplify some more

mvdbeek commented:
@natefoo I've added some changes in https://github.com/natefoo/galaxy/pull/10, with those the tests should pass

natefoo commented:
Thanks! I hadn't had time to get back to this.

natefoo committed:
Merge pull request #10 from mvdbeek/natefoo-installdbless-shed-tools

Simplify loading repositories and fix tests

mvdbeek commented:
@galaxybot test py3
